core/base/src/TObject.cxx

```diff
@@ -56,6 +56,69 @@ Bool_t TObject::fgObjectStat = kTRUE;
 
 ClassImp(TObject);
 
+namespace ROOT {
+namespace Internal {
+
+// Since we want to run the 'delete side effects' test early on,
+// we avoid using TClass/Cling so we have to use a mock class to
+// to estimate the offsets.
+struct MockTObject {
+   virtual ~MockTObject() = default;
+   // Int_t filler[10] = {0};
+   Int_t fBits = 0;
+   Int_t fUniqueID = 0;
+};
+
+// Return true if delete changes/poisoins/taints the memory.
+//
+// If operator delete tainted the memory, we can not rely on TestBit(kNotDeleted)
+// to check if the memory has been deleted (but in case, like TClonesArray,
+// where we know the destructor will be called but not operator delete, we can
+// still use to detect the cases where the destructor was called.
+
+bool DeleteChangesMemory()
+{
+   R__ASSERT(sizeof(MockTObject) == sizeof(TObject));
+   MockTObject m;
+   auto offset = ((char*)&(m.fBits)) - ((char*)&m); // Can not use `offsetof` because the virtual table make it a non-standard-layout type.
+
+   TObject *o = new TObject;
+   o->SetUniqueID(33);
+   unsigned char *oc = reinterpret_cast<unsigned char *>(o); // for address calculations
+   UInt_t *o_fuid = reinterpret_cast<UInt_t*>(oc + offset);
+   UInt_t *o_fbits = o_fuid + 1;
+   if (*o_fuid != 0x00000021) {
+      Error("CheckingDeleteSideEffects",
+            "fUniqueID is not at offset %zu as expected, we got 0x%.8x instead of 0x%.8x",
+            offset, *o_fuid, 0x00000021);
+   }
+   if (*o_fbits != 0x03000000) {
+      Error("CheckingDeleteSideEffects",
+            "fBits is not at offset %zu as expected, we got 0x%.8x instead of 0x%.8x",
+            offset + sizeof(UInt_t), *o_fbits, 0x03000000);
+   }
+   if (gDebug >= 9) {
+      unsigned char references[sizeof(TObject)];
+      memcpy(references, oc, sizeof(TObject));
+      delete o;
+      fprintf(stderr, "DEBUG: Checking before and after delete the content of a TObject with uniqueID 0x21");
+      for(size_t i = 0; i < sizeof(TObject); i += 4) {
+        fprintf(stderr, "DEBUG: 0x%.8x vs 0x%.8x\n", *(int*)(references +i), *(int*)(oc + i));
+      }
+   } else
+      delete o;
+
+   if ( *o_fbits != 0x01000000 ) {
```

Comment:
this access to `o_fbits` is technically UB, right? If so, I guess the assumption is that on all platforms where ROOT runs the implementation is such that, in practice, we can still safely access the memory. It might warrant a comment.

Reply:
Indeed.  This is the statement where we try to characterize the 'type' of UB we are seeing.